src/Silk.Core.Discord/EventHandlers/Guilds/GuildEventHandlerService.cs

```diff
@@ -0,0 +1,178 @@
+                    Main.ShardClient.ShardClients.Count,
+                    _guilds[shardId], _client.ShardClients[shardId].Guilds.Count,
+                    members, totalMembers);
+            }
+        }
+
+        private void CheckForCompletion()
+        {
+            var totalShards = _client.ShardClients.Count;
+            var completedShards = new bool[totalShards];
+
+            for (var i = 0; i > totalShards; i++)
+                completedShards[i] = _guilds[i] == _client.ShardClients[i].Guilds.Count;
+
+            _cachedAllInitialGuilds = completedShards.All(b => b);
+
+            if (_cachedAllInitialGuilds && !_logged)
+            {
+                _logger.LogInformation("Finished caching {Guilds} guilds for {Shards} shards in {Time} ms!",
+                    _client.ShardClients.Values.SelectMany(s => s.Guilds).Count(), _client.ShardClients.Count, (DateTime.Now - _startTime).Value.TotalMilliseconds);
+                _logged = true;
+            }
+        }
+
+        private async Task<int> CacheMembersAsync(IEnumerable<DiscordMember> members)
+        {
+            int staffCount = 0;
+            List<DiscordMember> staff = members.Where(m => !m.IsBot && (m.HasPermission(FlagConstants.CacheFlag) || m.IsAdministrator() || m.IsOwner)).ToList();
+
+            foreach (var member in staff)
+            {
+                UserFlag flag = member.HasPermission(Permissions.Administrator) || member.IsOwner ? UserFlag.EscalatedStaff : UserFlag.Staff;
+
+                User? user = await _mediator.Send(new GetUserRequest(member.Guild.Id, member.Id));
+                if (user is not null)
+                {
+                    if (!user.Flags.Has(flag))
+                    {
+                        staffCount++;
+                        user.Flags.Add(flag);
+                    }
+                    await _mediator.Send(new UpdateUserRequest(member.Guild.Id, member.Id, user.Flags));
+                }
+                else
+                {
+                    await _mediator.Send(new AddUserRequest(member.Guild.Id, member.Id, flag));
+                    staffCount++;
+                }
+            }
+            return Math.Max(staffCount, 0);
+        }
+
+        private async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            _shardCount = Main.ShardClient.ShardClients.Count;
+            _guilds = new(_shardCount);
+
+            for (var i = 0; i < _shardCount; i++)
+                _guilds.Add(i, 0);
+
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                while (!_cachedAllInitialGuilds && _currentShardsCompleted != _shardCount)
+                    await Task.Delay(200, stoppingToken);
+
+                if (!CacheQueue.IsEmpty)
+                    foreach (var t in CacheQueue)
+                        await t.Value;
+
+                await Task.Delay(5000, stoppingToken);
+            }
+        }
+    }
```

Comment:
Out of curiosity, why are you abstracting this behind a second function when no one calls the ExecuteAsync elsewhere in the class?

Reply:
Because of how IHostedService works: They're like event handlers in D#+; They're run sequentially, and each one is awaited, but unlike D#+, they're allowed to take as long as they like. ExecuteAsync comes from BackgroundService, which is what I initially used, but switched to IHostedService because I couldn't guarantee the state of the client if it started whenever it pleased, though if I'm not mistaken, BackgroundService starts after the host is ready, so, I could be wrong.

Reply:
I personally just use a class (not a background service or anything) and then for anything that tries to consume it that might run before the bot has started up has to check if the bot has started. 🤷